### ISSUE TITLE
Wishlist

### ISSUE THREAD

**Issue by sg3des** [tbogdala/fizzle#5]
Hello, i looked at several similar solutions, and can say - yuor engine is really good!

If you let, one remarks about imports and name packages: plz, do not use named imports ex:`graphics "github.com/tbogdala/fizzle/graphicsprovider"` - it makes impossible use autoimport tool (_goimports_), and second - your package name should be same with directory name - ex of problem: [forward_renderer](https://github.com/tbogdala/fizzle/blob/master/renderer/forward/forward_renderer.go), where name of package is **renderer** and name of directory is **forward**.

**Lights**:
1. direct light with direct shadows(sun).

**Objects**: 
1. Now resize(Scale) does not work for the imported objects;
2. How draw 2d graphics lines, circles and other.

**UI**, at least frames, buttons, progress bars.
Has several solutions creation UI on go with gl, ex: https://github.com/vizstra/ui, https://github.com/google/gxui... - but I could not combine them with fizzle. 

**Shaders**: How can I add own shaders? I mean so as not to change your engine code. i need:
1. shaders should implemented alpha-channel(blending);
2. shaders with colored glow, proceed from glow map;
3. if I'm not mistaken - prepared thee **diffuse_texbumped** shader does not take specular map;
4. animated textures from tex_00.png, text_01.png ... tex_NN.png for create water, and similars... 
5. texture tiling for generated mesh (Plane, Cube etc).

Most urgent is shader with alpha-channel and UI

THX.

**Comment by tbogdala** [comment]
>  it makes impossible use autoimport tool (goimports),

That I didn't know and I've been using named imports all over the place in all my projects.

> How draw 2d graphics lines, circles and other.

CreateWireframeCircleXZ creates a gl.LINES type of object for a circle on the XZ plane, but could be more customizable. In general, you are correct in that there isn't much 2d support.

> UI, at least frames, buttons, progress bars.

**Ah!** Now that I've been working on recently and intend on moving out the ui/font support from fizzle.
Look here: https://github.com/tbogdala/eweygewey

> Shaders: How can I add own shaders?

Just load them with LoadShaderProgramFromFiles() or LoadShaderProgram(). They can be from text files or string constants in your project.

re #3 i'm not sure i have any shaders that have specular maps right now.

There's one example of alpha blending, i think, in just the colortext shader.

**Comment by tbogdala** [comment]
Just as a separate comment, I want to add that I appreciate your feedback and will try to get to some of these issues.

I've been implementing things in Fizzle as I've needed them in whatever game prototype I'm working on and now that i've got my ui, networking, physics, graphics and now vr libraries coming along, I'll be focusing on getting a game together again. Hopefully during that time I can tighten up some of these shortcomings.

Thanks again for taking the time to write out a suggestion list!

**Comment by sg3des** [comment]
> >    UI, at least frames, buttons, progress bars.
> 
> Ah! Now that I've been working on recently and intend on moving out the ui/font support from fizzle.
> Look here: https://github.com/tbogdala/eweygewey

**Great!** I don`t know how can I miss it...

> >  Shaders: How can I add own shaders?
> 
> Just load them with LoadShaderProgramFromFiles() or LoadShaderProgram(). They can be from text files or string constants in your project.

i mean that all the variables used in shaders is hardcoded in https://github.com/tbogdala/fizzle/blob/master/renderer/renderer.go ... however, I am weak in shaders so this is yet not critical.

**Comment by sg3des** [comment]
> I've been implementing things in Fizzle as I've needed them in whatever game prototype I'm working on and now that i've got my ui, networking, physics, graphics and now vr libraries coming along, I'll be focusing on getting a game together again.

It is right! This will allow you to better understand the requirements of the real project. I've seen how people write engine for the engine - it is a dead-end direction. 

You probably already realized,  i am also working on a game-project using fizzle engine, coming soon will upload on github.

**Comment by tbogdala** [comment]
> i mean that all the variables used in shaders is hardcoded in https://github.com/tbogdala/fizzle/blob/master/renderer/renderer.go ... however, I am weak in shaders so this is yet not critical.

Yes, that's actually my strategy for now. They act kind of like built-ins for shaders for things you may want. Right now the generic bind & draw function will search to see if that variable is used and only bind content if it is used in the shader.

> I've seen how people write engine for the engine - it is a dead-end direction.

That's how I see it too. A little bit of that can help, but I find it best to ground the decisions in something practical. I'd be making more commits on fizzle if I had more free time, but for now it just gets updates when I need something.

> You probably already realized, i am also working on a game-project using fizzle engine, coming soon will upload on github.

That's **awesome**!

I'm about to do some more breaking changes by pulling out the 'ui' module and fonts.go and just have examples using eweygewey instead. I haven't made any version tags as of yet because things have been unstable. Do you think I should start having master for released tagged versions and then create a separate branch for development? Would that help out?

**Comment by tbogdala** [comment]
I have a UDP game networking library I haven't released yet and my sample game. I should probably upload them soon. I was waiting until I had just a little bit more on the game finished.

Also, I occasionally stream development on http://www.twitch.tv/animalmachine -- feel free to stop by and ask questions in real-time.

**Comment by sg3des** [comment]
> Now resize(Scale) does not work for the imported objects;

I was wrong - `Scale` effect on imported items too.

One question: Why all objects have first argument shader name (string) ? I can not see it`s affects.

One benchmark: I imported into the scene ~3 million faces, with transparency and shadows  - 200-300 FPS it`s amazing!

**Comment by tbogdala** [comment]
> One question: Why all objects have first argument shader name (string) ? I can not see it`s affects.

At one point in time, I intended this to facilitate a map lookup on a shader manager object. You are correct in that it isn't used. I should remove it.

> One benchmark: I imported into the scene ~3 million faces, with transparency and shadows - 200-300 FPS it`s amazing!

WooO!

Glad to hear its working for you.

**Comment by sg3des** [comment]
> >  One question: Why all objects have first argument shader name (string) ? I can not see it`s affects.
> 
> At one point in time, I intended this to facilitate a map lookup on a shader manager object. You are correct in that it isn't used. I should remove it.

I can help remove it, but this breaks a backward compatible - all inherited projects will need to be corrected...

**Comment by tbogdala** [comment]
As long as I agree with the direction of the API break, like I do with removing this, submit the patch with it removed and the samples fixed. The gold standard for Fizzle will be whether or not the examples function.  Eventually I'll start tagging versions.

Also, I've just started to use branches on some of my projects. The two that have `development` branches are fizzle and eweygewey. For this change, if you wanted to you could submit a patch against the development branch. I'm making a ton of API breaks right now, so its a good time to do it.

**Comment by sg3des** [comment]
Hi, I  see some changes in the engine, that`s great! 
I was away for a while... but now I'm going to continue work on my project, and as much as possible to help you with fizzle.

i`m uploaded on github my game-project - [sg3des/phasarv](https://github.com/sg3des/phasarv), which is based on your engine. Please see it, if you're interested.

about wishlist, there was one point - **"direct light with direct shadows(sun)"**

I've seen the changes in texturing objects, why you do not want to realize the likeness of material? 
masterial is struct contains shader and all necessary textures, example:

```
type Material struct {
  Shader *fizzle.RenderShader 
  Textures map[string]graphicsprovider.Texture
  Colors map[string]mgl32.Vec4
  [...]
}
```

maybe a little differently... 

where in Textures map - `Textures = map[string]graphicsprovider.Texture{"MATERIAL_TEX_0": tex0, "MATERIAL_TEX_1":tex1,...}`
in Colors map - `Colors = map[string]mgl32.Vec4{"MATERIAL_DIFFUSE":mgl32.Vec4{1,0,0,1}...}`

this makes it possible use one material for multiple objects, 
more comfortable creation of objects, 
use any name of keys in shaders - not hardcoded like now.

**Comment by tbogdala** [comment]
Good to hear from you again @sg3des! I've got tagged releases now and development is happening in the devlopment branch of the repository, so things are starting to get more organized. Many of my repositories now have a development branch, actually. Look for the CHANGELOG.md files to catch up.

As for your comments about materials, the answer is yes, that'd be probably the right thing to do for the reasons you mentioned. It would also allow an optimization for sequential draw calls using the same material to not have to make certain GL calls like using shader, setting material uniforms, etc ... since they'd be the same value.

I don't have a good reason not to use them. They're like this now because that's how I hacked it together from the prototype stages.

---

I've also been thinking about pulling in the common shaders I use (basic, basicSkinned, color and colorText) into the source code itself so that you can create a shader with one function call and not have to load them from file. It'd also allow me to make custom uniform/attribute binders to make it a bit quicker.

I'll probably get to this stuff soon because it's been on my brain as well. Within the calendar week. It's time.

**Comment by tbogdala** [comment]
Updated development branch with a new Material structure. Next step is to bundle in the standard shaders as I now have at least 3 different instances of them and it's causing me problems.

Edit: the basic set of shaders is now built into the `renderer/forward` package.
